cellengine/resources/compensation.py

```diff
@@ -1,68 +1,97 @@
 from __future__ import annotations
-from dataclasses import dataclass, field
 from typing import List, TYPE_CHECKING
 
-from dataclasses_json.cfg import config
-import numpy
+from attr import define, field
+from numpy import array, linalg
 from pandas import DataFrame
 
 import cellengine as ce
-from cellengine.utils.dataclass_mixin import DataClassMixin, ReadOnly
+from cellengine.utils import readonly
+from cellengine.utils.converter import converter
 
 
 if TYPE_CHECKING:
     from cellengine.resources.fcs_file import FcsFile
 
 
-@dataclass
-class Compensation(DataClassMixin):
+@define
+class Compensation:
     """A class representing a CellEngine compensation matrix.
 
     Can be applied to FCS files to compensate them.
     """
 
+    _id: str = field(on_setattr=readonly)
+    experiment_id: str = field(on_setattr=readonly)
     name: str
     channels: List[str]
-    dataframe: DataFrame = field(
-        metadata=config(
-            field_name="spillMatrix",
-            encoder=lambda x: x.to_numpy().flatten().tolist(),
-            decoder=lambda x: numpy.array(x),
-        )
-    )
-    _id: str = field(
-        metadata=config(field_name="_id"), default=ReadOnly()
-    )  # type: ignore
-    experiment_id: str = field(default=ReadOnly())  # type: ignore
-
-    def __post_init__(self):
-        self.dataframe = DataFrame(
-            self.dataframe.reshape(self.N, self.N),
+    spill_matrix: List[float]
+
+    @property
+    def dataframe(self):
+        return DataFrame(
+            array(self.spill_matrix).reshape(self.N, self.N),  # type: ignore
             columns=self.channels,
             index=self.channels,
         )
 
+    @dataframe.setter
+    def dataframe(self, val: DataFrame):
+        try:
+            assert len(val.columns) == len(val.index)
```

Comment:
I think this line is redundant with the next one.
```
>>> df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]}, index=["a", "b"])                     
>>> assert all(df.columns == df.index)
...
ValueError: Lengths must match to compare
```

Reply:
:+1: